File: src/test/java/io/protobase/f7/formulas/math/MULTIPLYTest.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.NAException;
import io.protobase.f7.errors.ValueException;
import io.protobase.f7.models.Grid;
import io.protobase.f7.testutils.TestFormula;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class MULTIPLYTest extends TestFormula {
  @Test
  public void testApply_cleanDivision() {
    assertThat(MULTIPLY.SELF.apply(null, 10.0, 10.0)).isEqualTo(100.0);
    assertThat(MULTIPLY.SELF.apply(null, 10.0, 2.0)).isEqualTo(20.0);
    assertThat(MULTIPLY.SELF.apply(null, 10.0, 3.0)).isEqualTo(30.0);
    assertThat(MULTIPLY.SELF.apply(null, 0.0, 1628736813.2)).isEqualTo(0.0);
    assertThat(MULTIPLY.SELF.apply(null, 21.22, 2876.111)).isEqualTo(61031.075419999994);
  }

  @Test
  public void testApply_stringConversion() {
    assertThat(MULTIPLY.SELF.apply(null, "10.0", "10.0")).isEqualTo(100.0);
  }

  @Test
  public void testApply_grid() {
    assertThat(MULTIPLY.SELF.apply(null,
        Grid.builder().add(0, 0, 6.0).add(0, 1, "Don't mind me.").build(),
        Grid.builder().add(0, 0, 2.0).add(0, 1, "Don't mind me.").build()
    )).isEqualTo(12.0);
  }

  @Test
  public void test_lookup() {
    MULTIPLY F = new MULTIPLY(lookup, collateralLookup);
    when(collateralLookup.apply(A1, M22_RANGE)).thenReturn(2.0);
    when(collateralLookup.apply(A1, G19_RANGE)).thenReturn(10.0);
    assertThat(F.apply(A1, M22_RANGE, G19_RANGE)).isEqualTo(20.0);
    verify(collateralLookup).apply(A1, M22_RANGE);
    verify(collateralLookup).apply(A1, G19_RANGE);
    verifyNoMoreInteractions(collateralLookup);
  }

  @Test
  public void testApply_errorsPassThrough() {
    assertThat(MULTIPLY.SELF.apply(null, 10.0, new ValueException())).isEqualTo(new ValueException());
    assertThat(MULTIPLY.SELF.apply(null, new ValueException(), 10.0)).isEqualTo(new ValueException());
  }

  @Test
  public void testApply_argumentsMismatch() {
    assertThat(MULTIPLY.SELF.apply(null, "Too few")).isEqualTo(new NAException());
    assertThat(MULTIPLY.SELF.apply(null, "A", "B", "Too many")).isEqualTo(new NAException());
  }
}

File: src/test/js/formulas/math/MULTIPLYTest.ts
import { assert } from "chai";
import { it, describe } from "../../testutils/TestUtils";
import { stub } from "sinon";
import { F7Exception } from "../../../../main/js/errors/F7Exception";
import { F7ExceptionName } from "../../../../main/js/errors/F7ExceptionName";
import { NAException } from "../../../../main/js/errors/NAException";
import { ValueException } from "../../../../main/js/errors/ValueException";
import { MULTIPLY } from "../../../../main/js/formulas/math/MULTIPLY";
import { Grid } from "../../../../main/js/models/common/Grid";
import { CollateralLookupFunction, LookupFunction } from "../../../../main/js/models/common/Types";
import { CommonModels } from "../../testutils/CommonModels";

describe("MULTIPLY", function () {
  it("should do normal operations", function () {
    assert.deepEqual(MULTIPLY.SELF.run(null, 10, 10), 100);
    assert.deepEqual(MULTIPLY.SELF.run(null, 10, 2), 20);
    assert.deepEqual(MULTIPLY.SELF.run(null, 10, 3), 30);
    assert.deepEqual(MULTIPLY.SELF.run(null, 0, 1628736813.2), 0);
    assert.deepEqual(MULTIPLY.SELF.run(null, 21.22, 2876.111), 61031.075419999994);
  });

  it("should do string conversion", function () {
    assert.deepEqual(MULTIPLY.SELF.run(null, "10", "10"), 100);
    assert.deepEqual(MULTIPLY.SELF.run(null, "10", "2"), 20);
    assert.deepEqual(MULTIPLY.SELF.run(null, "10", "3"), 30);
  });

  it("should do pass-through errors", function () {
    assert.deepEqual(
      (MULTIPLY.SELF.run(null, 10, new ValueException()) as F7Exception).name,
      F7ExceptionName.VALUE
    );
    assert.deepEqual(
      (MULTIPLY.SELF.run(null, new ValueException(), 10) as F7Exception).name,
      F7ExceptionName.VALUE
    );
  });

  it("should use lookup", function () {
    const lookup = stub();
    const collateralLookup = stub();
    const F = new MULTIPLY(lookup as LookupFunction, collateralLookup as CollateralLookupFunction);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.M22_RANGE).returns(6);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.G19_RANGE).returns(2);
    assert.deepEqual(F.run(CommonModels.A1, CommonModels.M22_RANGE, CommonModels.G19_RANGE), 12);
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.M22_RANGE));
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.G19_RANGE));
    assert.equal(collateralLookup.callCount, 2);
    assert.isTrue(lookup.notCalled);
  });

  it("should handle grids", function () {
    const one = Grid.builder().add(0, 0, 6).add(0, 1, "A").build();
    const two = Grid.builder().add(0, 0, 2).add(0, 1, "B").build();
    assert.deepEqual(MULTIPLY.SELF.run(null, one, two), 12);
  });

  it("should return error when argument lengths are wrong", function () {
    assert.deepEqual((MULTIPLY.SELF.run(null, "Too few") as NAException).name, F7ExceptionName.NA);
    assert.deepEqual(
      (MULTIPLY.SELF.run(null, "A", "B", "Too many") as NAException).name,
      F7ExceptionName.NA
    );
  });
});
